src/library/scala/concurrent/impl/Promise.scala

```diff
@@ -178,7 +178,9 @@ private[concurrent] object Promise {
    * DefaultPromises, and `linkedRootOf` is currently only designed to be called
    * by Future.flatMap.
    */
-  final class DefaultPromise[T] extends AtomicReference[AnyRef](Nil) with Promise[T] {
+  // Left non-final to enable addition of extra fields by Java/Scala converters
+  // in scala-java8-compat.
+  class DefaultPromise[T] extends AtomicReference[AnyRef](Nil) with Promise[T] {
```

Comment:
Yes, so AtomicReferenceFieldUpdater will not save us any memory here, and will lose us performance.

If we want to opt for memory usage we should optimize the structure of callbacks. I've got a couple of ideas but no time to work on that at the moment.